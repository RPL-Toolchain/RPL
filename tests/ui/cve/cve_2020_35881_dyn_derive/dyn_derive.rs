pub trait Fat<T: ?Sized>: AsRef<T> + Sized {
    fn into_box(self, f: impl FnOnce(Self) -> *mut ()) -> Box<T> {
        let mut fat_ptr = self.as_ref() as *const T;
        let data_ptr = &mut fat_ptr as *mut *const T as *mut *mut ();
        //~^ ERROR: wrong assumption of fat pointer layout
        unsafe {
            *data_ptr = f(self);
            Box::from_raw(fat_ptr as *mut T)
        }
    }

    fn to_box(self, f: impl FnOnce(&T) -> *mut ()) -> Box<T> {
        let mut fat_ptr = self.as_ref() as *const T;
        let data_ptr = &mut fat_ptr as *mut *const T as *mut *mut ();
        //~^ ERROR: wrong assumption of fat pointer layout
        unsafe {
            *data_ptr = f(self.as_ref());
            Box::from_raw(fat_ptr as *mut T)
        }
    }
}
