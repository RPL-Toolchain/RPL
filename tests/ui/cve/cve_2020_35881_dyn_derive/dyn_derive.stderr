error: wrong assumption of fat pointer layout
  --> tests/ui/cve/cve_2020_35881_dyn_derive/dyn_derive.rs:4:24
   |
LL |         let data_ptr = &mut fat_ptr as *mut *const T as *mut *mut ();
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ptr transmute here
...
LL |             *data_ptr = f(self);
   |                         ------- try to get data ptr from first 8 bytes here
   |
   = help: the Rust Compiler does not expose the layout of fat pointers
   = note: `#[deny(rpl::wrong_assumption_of_fat_pointer_layout)]` on by default

error: wrong assumption of fat pointer layout
  --> tests/ui/cve/cve_2020_35881_dyn_derive/dyn_derive.rs:14:24
   |
LL |         let data_ptr = &mut fat_ptr as *mut *const T as *mut *mut ();
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ptr transmute here
...
LL |             *data_ptr = f(self.as_ref());
   |                         ---------------- try to get data ptr from first 8 bytes here
   |
   = help: the Rust Compiler does not expose the layout of fat pointers

error: aborting due to 2 previous errors

